confignode/src/main/java/org/apache/iotdb/confignode/manager/node/NodeManager.java

```diff
@@ -255,8 +255,17 @@ public DataSet removeDataNode(RemoveDataNodePlan removeDataNodePlan) {
           preCheckStatus.getStatus());
       return preCheckStatus;
     }
-    // if add request to queue, then return to client
+
     DataNodeToStatusResp dataSet = new DataNodeToStatusResp();
+    // do transfer of the DataNodes before remove
+    if (configManager.transfer(removeDataNodePlan.getDataNodeLocations()).getCode()
+        != TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+      dataSet.setStatus(
+          new TSStatus(TSStatusCode.NODE_DELETE_FAILED_ERROR.getStatusCode())
+              .setMessage("Fail to do transfer of the DataNodes"));
+      return dataSet;
+    }
```

Comment:
Is the `transfer` process should be such important that the DataNode remove process couldn't continue as long as if it fails down?

Reply:
Yes, I think we should ensure the safety of data.

Reply:
Remove data process used RemoveDataProcedure and sub RegionMigrateProcedures, 
Will triggers implement it later or using other methods?
Or some errors may happen when ConfigNode down/restart.

![image](https://user-images.githubusercontent.com/6756545/196654189-c92c66a5-3dd8-42ed-aafc-717d206de379.png)

Reply:
We used to consider implement by Procedure, but later we found that we did not need to roll back when the transfer failed.